File: hash-map-set-table-(hashing)/2131.py
# Tc = O(n) [number of words]
# Sc = O(n)
from collections import defaultdict
class Solution:
    def longestPalindrome(self, words):
        mp = defaultdict(int)
        result = 0
        for word in words:
            reversed_word = word[::-1]
            if mp[reversed_word] > 0:
                result += 4
                mp[reversed_word] -= 1
            else:
                mp[word] += 1
        
        # check equal character words and use only once
        for word, count in mp.items():
            if word[0] == word[1] and count > 0:
                result += 2
                break
        return result
            

File: hash-map-set-table-(hashing)/2131.js
//  Tc = O(n) [number of words]
//  Sc = O(n)
class Solution {
    longestPalindrome(words) {
        const mp = new Map();
        let result = 0;
        
        for (const word of words) {
            const reversedWord = word.split('').reverse().join('');
            if (mp.get(reversedWord) > 0) {
                result += 4;
                mp.set(reversedWord, mp.get(reversedWord) - 1);
            } else {
                mp.set(word, (mp.get(word) || 0) + 1);
            }
        }
        
        // check equal character words and use only once
        for (const [word, count] of mp.entries()) {
            if (word[0] === word[1] && count > 0) {
                result += 2;
                break;
            }
        }
        return result;
    }
}